

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>scribe.svi &mdash; SCRIBE  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            SCRIBE
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickoverview.html">Quick Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Available Models</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../models/index.html">Models</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../results.html">Results Class</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/index.html">Examples</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">SCRIBE</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">scribe.svi</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for scribe.svi</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Stochastic Variational Inference (SVI) module for single-cell RNA sequencing</span>
<span class="sd">data analysis.</span>

<span class="sd">This module implements a Dirichlet-Multinomial model for scRNA-seq count data</span>
<span class="sd">using Numpyro for variational inference.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># Imports for inference</span>
<span class="kn">from</span> <span class="nn">jax</span> <span class="kn">import</span> <span class="n">random</span>
<span class="kn">import</span> <span class="nn">jax.numpy</span> <span class="k">as</span> <span class="nn">jnp</span>
<span class="kn">import</span> <span class="nn">numpyro</span>
<span class="kn">import</span> <span class="nn">numpyro.distributions</span> <span class="k">as</span> <span class="nn">dist</span>
<span class="kn">from</span> <span class="nn">numpyro.infer</span> <span class="kn">import</span> <span class="n">SVI</span><span class="p">,</span> <span class="n">TraceMeanField_ELBO</span>

<span class="c1"># Imports for results</span>
<span class="kn">from</span> <span class="nn">.results</span> <span class="kn">import</span> <span class="o">*</span>

<span class="c1"># Imports for data handling</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">import</span> <span class="nn">scipy.sparse</span>

<span class="c1"># Imports for AnnData</span>
<span class="kn">from</span> <span class="nn">anndata</span> <span class="kn">import</span> <span class="n">AnnData</span>

<span class="c1"># Imports for model-specific functions</span>
<span class="kn">from</span> <span class="nn">.model_registry</span> <span class="kn">import</span> <span class="n">get_model_and_guide</span>

<span class="c1"># ------------------------------------------------------------------------------</span>
<span class="c1"># Stochastic Variational Inference with Numpyro</span>
<span class="c1"># ------------------------------------------------------------------------------</span>

<div class="viewcode-block" id="create_svi_instance">
<a class="viewcode-back" href="../../api/svi.html#scribe.svi.create_svi_instance">[docs]</a>
<span class="k">def</span> <span class="nf">create_svi_instance</span><span class="p">(</span>
    <span class="n">model_type</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">optimizer</span><span class="p">:</span> <span class="n">numpyro</span><span class="o">.</span><span class="n">optim</span><span class="o">.</span><span class="n">optimizers</span> <span class="o">=</span> <span class="n">numpyro</span><span class="o">.</span><span class="n">optim</span><span class="o">.</span><span class="n">Adam</span><span class="p">(</span><span class="n">step_size</span><span class="o">=</span><span class="mf">0.001</span><span class="p">),</span>
    <span class="n">loss</span><span class="p">:</span> <span class="n">numpyro</span><span class="o">.</span><span class="n">infer</span><span class="o">.</span><span class="n">elbo</span> <span class="o">=</span> <span class="n">TraceMeanField_ELBO</span><span class="p">(),</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create an SVI instance with the defined model and guide.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    model_type : str, optional</span>
<span class="sd">        Type of model to use. Must be one of:</span>
<span class="sd">            - &quot;nbdm&quot;: Negative Binomial model</span>
<span class="sd">            - &quot;zinb&quot;: Zero-Inflated Negative Binomial model</span>
<span class="sd">            - &quot;nbvcp&quot;: Negative Binomial with variable capture probability</span>
<span class="sd">            - &quot;zinbvcp&quot;: Zero-Inflated Negative Binomial with variable capture</span>
<span class="sd">              probability</span>
<span class="sd">            - Mixture variants with &quot;_mix&quot; suffix (e.g. &quot;nbdm_mix&quot;)</span>
<span class="sd">    optimizer : numpyro.optim.optimizers, optional</span>
<span class="sd">        Optimizer to use for stochastic optimization. Default is Adam with </span>
<span class="sd">        step_size=0.001</span>
<span class="sd">    loss : numpyro.infer.elbo, optional</span>
<span class="sd">        Loss function for variational inference. Default is TraceMeanField_ELBO()</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    numpyro.infer.SVI</span>
<span class="sd">        Configured SVI instance ready for training</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">        If model_type is invalid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get model and guide functions</span>
    <span class="n">model_fn</span><span class="p">,</span> <span class="n">guide_fn</span> <span class="o">=</span> <span class="n">get_model_and_guide</span><span class="p">(</span><span class="n">model_type</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">SVI</span><span class="p">(</span>
        <span class="n">model_fn</span><span class="p">,</span>
        <span class="n">guide_fn</span><span class="p">,</span>
        <span class="n">optimizer</span><span class="p">,</span>
        <span class="n">loss</span><span class="o">=</span><span class="n">loss</span>
    <span class="p">)</span></div>


<span class="c1"># ------------------------------------------------------------------------------</span>

<div class="viewcode-block" id="run_inference">
<a class="viewcode-back" href="../../api/svi.html#scribe.svi.run_inference">[docs]</a>
<span class="k">def</span> <span class="nf">run_inference</span><span class="p">(</span>
    <span class="n">svi_instance</span><span class="p">:</span> <span class="n">SVI</span><span class="p">,</span>
    <span class="n">rng_key</span><span class="p">:</span> <span class="n">random</span><span class="o">.</span><span class="n">PRNGKey</span><span class="p">,</span>
    <span class="n">counts</span><span class="p">:</span> <span class="n">jnp</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">model_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;nbdm&quot;</span><span class="p">,</span>
    <span class="n">model_config</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">ModelConfig</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">n_steps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100_000</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">cells_axis</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">stable_update</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpyro</span><span class="o">.</span><span class="n">infer</span><span class="o">.</span><span class="n">svi</span><span class="o">.</span><span class="n">SVIRunResult</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run stochastic variational inference on the provided count data.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    svi_instance : numpyro.infer.SVI</span>
<span class="sd">        Configured SVI instance for running inference. NOTE: Make sure that this</span>
<span class="sd">        was created with the same model and guide as the one being used here!</span>
<span class="sd">    rng_key : jax.random.PRNGKey</span>
<span class="sd">        Random number generator key</span>
<span class="sd">    counts : jax.numpy.ndarray</span>
<span class="sd">        Count matrix. If cells_axis=0 (default), shape is (n_cells, n_genes).</span>
<span class="sd">        If cells_axis=1, shape is (n_genes, n_cells).</span>
<span class="sd">    n_steps : int, optional</span>
<span class="sd">        Number of optimization steps. Default is 100,000</span>
<span class="sd">    batch_size : int, optional</span>
<span class="sd">        Mini-batch size for stochastic optimization. Default is 512</span>
<span class="sd">    model_type : str, optional</span>
<span class="sd">        Type of model being used. Built-in options are:</span>
<span class="sd">        - &quot;nbdm&quot;: Negative Binomial-Dirichlet Multinomial model</span>
<span class="sd">        - &quot;zinb&quot;: Zero-Inflated Negative Binomial model</span>
<span class="sd">        - &quot;nbvcp&quot;: Negative Binomial with variable capture probability</span>
<span class="sd">        - &quot;zinbvcp&quot;: Zero-Inflated Negative Binomial with variable capture</span>
<span class="sd">          probability</span>
<span class="sd">        - &quot;nbdm_mix&quot;: Negative Binomial-Dirichlet Multinomial Mixture Model</span>
<span class="sd">    n_components : int, optional</span>
<span class="sd">        Number of components to fit for mixture models. Default is None.</span>
<span class="sd">    prior_params : Dict, optional</span>
<span class="sd">        Dictionary of prior parameters specific to the model.</span>
<span class="sd">    cells_axis : int, optional</span>
<span class="sd">        Axis along which cells are arranged. 0 means cells are rows (default), 1</span>
<span class="sd">        means cells are columns.</span>
<span class="sd">    custom_args : Dict, optional</span>
<span class="sd">        Dictionary of additional arguments to pass to any custom model.</span>
<span class="sd">    stable_update : bool, optional</span>
<span class="sd">        Whether to use stable update method. Default is True. If true,  returns</span>
<span class="sd">        the current state if the the loss or the new state contains invalid</span>
<span class="sd">        values.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    numpyro.infer.svi.SVIRunResult</span>
<span class="sd">        Results from the SVI run containing optimized parameters and loss</span>
<span class="sd">        history</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract dimensions and compute total counts based on cells axis</span>
    <span class="k">if</span> <span class="n">cells_axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">n_cells</span><span class="p">,</span> <span class="n">n_genes</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">shape</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_cells</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">shape</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># Transpose to make cells rows for model</span>

    <span class="c1"># Prepare base arguments that all models should receive</span>
    <span class="n">model_args</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;n_cells&#39;</span><span class="p">:</span> <span class="n">n_cells</span><span class="p">,</span>
        <span class="s1">&#39;n_genes&#39;</span><span class="p">:</span> <span class="n">n_genes</span><span class="p">,</span>
        <span class="s1">&#39;counts&#39;</span><span class="p">:</span> <span class="n">counts</span><span class="p">,</span>
        <span class="s1">&#39;batch_size&#39;</span><span class="p">:</span> <span class="n">batch_size</span><span class="p">,</span>
        <span class="s1">&#39;model_config&#39;</span><span class="p">:</span> <span class="n">model_config</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="c1"># Add model-specific parameters</span>
    <span class="k">if</span> <span class="n">model_type</span> <span class="o">==</span> <span class="s2">&quot;nbdm&quot;</span><span class="p">:</span>
        <span class="n">model_args</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Check if mixture model</span>
    <span class="k">if</span> <span class="n">model_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="s2">&quot;mix&quot;</span> <span class="ow">in</span> <span class="n">model_type</span><span class="p">:</span>
        <span class="c1"># Check if n_components is provided</span>
        <span class="k">if</span> <span class="n">model_config</span><span class="o">.</span><span class="n">n_components</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;n_components must be specified for mixture model </span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">model_config</span><span class="o">.</span><span class="n">n_components</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;n_components must be greater than 1 for mixture model </span><span class="si">{</span><span class="n">model_type</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

    <span class="c1"># Run the inference algorithm</span>
    <span class="k">return</span> <span class="n">svi_instance</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="n">rng_key</span><span class="p">,</span>
        <span class="n">n_steps</span><span class="p">,</span>
        <span class="n">stable_update</span><span class="o">=</span><span class="n">stable_update</span><span class="p">,</span>
        <span class="o">**</span><span class="n">model_args</span>
    <span class="p">)</span></div>


<span class="c1"># ------------------------------------------------------------------------------</span>
<span class="c1"># SCRIBE inference pipeline</span>
<span class="c1"># ------------------------------------------------------------------------------</span>

<div class="viewcode-block" id="run_scribe">
<a class="viewcode-back" href="../../api/svi.html#scribe.svi.run_scribe">[docs]</a>
<span class="k">def</span> <span class="nf">run_scribe</span><span class="p">(</span>
    <span class="n">counts</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">jnp</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="s2">&quot;AnnData&quot;</span><span class="p">],</span>
    <span class="c1"># Model configuration</span>
    <span class="n">zero_inflated</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">variable_capture</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> 
    <span class="n">mixture_model</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">n_components</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="c1"># Distribution choices</span>
    <span class="n">r_dist</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gamma&quot;</span><span class="p">,</span>
    <span class="c1"># Prior parameters</span>
    <span class="n">r_prior</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">p_prior</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">gate_prior</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">p_capture_prior</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">mixing_prior</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="c1"># Training parameters</span>
    <span class="n">n_steps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100_000</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">optimizer</span><span class="p">:</span> <span class="n">numpyro</span><span class="o">.</span><span class="n">optim</span><span class="o">.</span><span class="n">optimizers</span> <span class="o">=</span> <span class="n">numpyro</span><span class="o">.</span><span class="n">optim</span><span class="o">.</span><span class="n">Adam</span><span class="p">(</span><span class="n">step_size</span><span class="o">=</span><span class="mf">0.001</span><span class="p">),</span>
    <span class="c1"># Data handling options</span>
    <span class="n">cells_axis</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">layer</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="c1"># Extra options</span>
    <span class="n">seed</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">42</span><span class="p">,</span>
    <span class="n">stable_update</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">r_guide</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">loss</span><span class="p">:</span> <span class="n">numpyro</span><span class="o">.</span><span class="n">infer</span><span class="o">.</span><span class="n">elbo</span> <span class="o">=</span> <span class="n">TraceMeanField_ELBO</span><span class="p">(),</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ScribeResults</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run SCRIBE inference pipeline to fit a probabilistic model to single-cell</span>
<span class="sd">    RNA sequencing data.</span>
<span class="sd">    </span>
<span class="sd">    This function provides a high-level interface to configure and run SCRIBE</span>
<span class="sd">    models. It handles data preprocessing, model configuration, variational</span>
<span class="sd">    inference, and results packaging.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    counts : Union[jnp.ndarray, AnnData]</span>
<span class="sd">        Count matrix or AnnData object containing counts. If AnnData, counts</span>
<span class="sd">        should be in .X or specified layer. Shape should be (cells, genes) if</span>
<span class="sd">        cells_axis=0, or (genes, cells) if cells_axis=1.</span>
<span class="sd">        </span>
<span class="sd">    Model Configuration:</span>
<span class="sd">    -------------------</span>
<span class="sd">    zero_inflated : bool, default=False</span>
<span class="sd">        Whether to use zero-inflated negative binomial (ZINB) model to account</span>
<span class="sd">        for dropout events</span>
<span class="sd">    variable_capture : bool, default=False </span>
<span class="sd">        Whether to model cell-specific mRNA capture efficiencies</span>
<span class="sd">    mixture_model : bool, default=False</span>
<span class="sd">        Whether to use mixture model components for heterogeneous populations</span>
<span class="sd">    n_components : Optional[int], default=None</span>
<span class="sd">        Number of mixture components. Required if mixture_model=True.</span>
<span class="sd">        </span>
<span class="sd">    Distribution Choices:</span>
<span class="sd">    -------------------</span>
<span class="sd">    r_dist : str, default=&quot;gamma&quot;</span>
<span class="sd">        Distribution family for dispersion parameter. Options:</span>
<span class="sd">            - &quot;gamma&quot;: Gamma distribution (default)</span>
<span class="sd">            - &quot;lognormal&quot;: Log-normal distribution</span>
<span class="sd">        </span>
<span class="sd">    Prior Parameters:</span>
<span class="sd">    ----------------</span>
<span class="sd">    r_prior : Optional[tuple], default=None</span>
<span class="sd">        Prior parameters for dispersion (r) distribution:</span>
<span class="sd">            - For gamma: (shape, rate), defaults to (2, 0.1)</span>
<span class="sd">            - For lognormal: (mu, sigma), defaults to (1, 1)</span>
<span class="sd">    p_prior : Optional[tuple], default=None</span>
<span class="sd">        Prior parameters (alpha, beta) for success probability Beta</span>
<span class="sd">        distribution. Defaults to (1, 1).</span>
<span class="sd">    gate_prior : Optional[tuple], default=None</span>
<span class="sd">        Prior parameters (alpha, beta) for dropout gate Beta distribution. Only</span>
<span class="sd">        used if zero_inflated=True. Defaults to (1, 1).</span>
<span class="sd">    p_capture_prior : Optional[tuple], default=None</span>
<span class="sd">        Prior parameters (alpha, beta) for capture efficiency Beta distribution.</span>
<span class="sd">        Only used if variable_capture=True. Defaults to (1, 1).</span>
<span class="sd">    mixing_prior : Optional[tuple], default=None</span>
<span class="sd">        Prior concentration parameter(s) for mixture weights Dirichlet</span>
<span class="sd">        distribution. Required if mixture_model=True.</span>
<span class="sd">        </span>
<span class="sd">    Training Parameters:</span>
<span class="sd">    ------------------</span>
<span class="sd">    n_steps : int, default=100_000</span>
<span class="sd">        Number of optimization steps for variational inference</span>
<span class="sd">    batch_size : Optional[int], default=None</span>
<span class="sd">        Mini-batch size for stochastic optimization. If None, uses full dataset.</span>
<span class="sd">    optimizer : numpyro.optim.optimizers, default=Adam(step_size=0.001)</span>
<span class="sd">        Optimizer for variational inference</span>
<span class="sd">        </span>
<span class="sd">    Data Handling:</span>
<span class="sd">    -------------</span>
<span class="sd">    cells_axis : int, default=0</span>
<span class="sd">        Axis for cells in count matrix (0=rows, 1=columns)</span>
<span class="sd">    layer : Optional[str], default=None</span>
<span class="sd">        Layer in AnnData to use for counts. If None, uses .X.</span>
<span class="sd">        </span>
<span class="sd">    Additional Options:</span>
<span class="sd">    -----------------</span>
<span class="sd">    seed : int, default=42</span>
<span class="sd">        Random seed for reproducibility</span>
<span class="sd">    stable_update : bool, default=True</span>
<span class="sd">        Whether to use numerically stable parameter updates during optimization</span>
<span class="sd">    r_guide : Optional[str], default=None</span>
<span class="sd">        Distribution family for guide of dispersion parameter. If None, uses</span>
<span class="sd">        same as r_dist. Options: &quot;gamma&quot; or &quot;lognormal&quot;.</span>
<span class="sd">    loss : numpyro.infer.elbo, default=TraceMeanField_ELBO()</span>
<span class="sd">        Loss function for variational inference</span>
<span class="sd">        </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    ScribeResults</span>
<span class="sd">        Results object containing:</span>
<span class="sd">            - Fitted model parameters</span>
<span class="sd">            - Loss history</span>
<span class="sd">            - Model configuration</span>
<span class="sd">            - Prior parameters</span>
<span class="sd">            - Dataset metadata</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Determine model type based on boolean flags</span>
    <span class="n">base_model</span> <span class="o">=</span> <span class="s2">&quot;nbdm&quot;</span>
    <span class="k">if</span> <span class="n">zero_inflated</span><span class="p">:</span>
        <span class="n">base_model</span> <span class="o">=</span> <span class="s2">&quot;zinb&quot;</span>
    <span class="k">if</span> <span class="n">variable_capture</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">zero_inflated</span><span class="p">:</span>
            <span class="n">base_model</span> <span class="o">=</span> <span class="s2">&quot;zinbvcp&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">base_model</span> <span class="o">=</span> <span class="s2">&quot;nbvcp&quot;</span>
    
    <span class="k">if</span> <span class="n">n_components</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">mixture_model</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;n_components must be None when mixture_model=False&quot;</span>
        <span class="p">)</span>
            
    <span class="c1"># Define model type</span>
    <span class="n">model_type</span> <span class="o">=</span> <span class="n">base_model</span>
    <span class="c1"># Check if mixture model</span>
    <span class="k">if</span> <span class="n">mixture_model</span><span class="p">:</span>
        <span class="c1"># Validate n_components for mixture models</span>
        <span class="k">if</span> <span class="n">n_components</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">n_components</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;n_components must be specified and greater than 1 &quot;</span>
                <span class="s2">&quot;when mixture_model=True&quot;</span>
            <span class="p">)</span>
        <span class="c1"># Validate mixing_prior for mixture models</span>
        <span class="k">if</span> <span class="n">mixing_prior</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;mixing_prior must be specified when mixture_model=True&quot;</span>
            <span class="p">)</span>
        <span class="c1"># Add mixture suffix if needed</span>
        <span class="n">model_type</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">base_model</span><span class="si">}</span><span class="s2">_mix&quot;</span>
    
    <span class="c1"># Handle AnnData input</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="s2">&quot;obs&quot;</span><span class="p">):</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">counts</span>
        <span class="n">count_data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">layer</span><span class="p">]</span> <span class="k">if</span> <span class="n">layer</span> <span class="k">else</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
        <span class="k">if</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">issparse</span><span class="p">(</span><span class="n">count_data</span><span class="p">):</span>
            <span class="n">count_data</span> <span class="o">=</span> <span class="n">count_data</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
        <span class="n">count_data</span> <span class="o">=</span> <span class="n">jnp</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">count_data</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">count_data</span> <span class="o">=</span> <span class="n">counts</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Extract dimensions</span>
    <span class="k">if</span> <span class="n">cells_axis</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">n_cells</span><span class="p">,</span> <span class="n">n_genes</span> <span class="o">=</span> <span class="n">count_data</span><span class="o">.</span><span class="n">shape</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">n_genes</span><span class="p">,</span> <span class="n">n_cells</span> <span class="o">=</span> <span class="n">count_data</span><span class="o">.</span><span class="n">shape</span>
        <span class="n">count_data</span> <span class="o">=</span> <span class="n">count_data</span><span class="o">.</span><span class="n">T</span>
    
    <span class="c1"># Set default priors based on distribution choices</span>
    <span class="k">if</span> <span class="n">r_prior</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">r_prior</span> <span class="o">=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span> <span class="k">if</span> <span class="n">r_dist</span> <span class="o">==</span> <span class="s2">&quot;gamma&quot;</span> <span class="k">else</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">p_prior</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">p_prior</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;zinb&quot;</span> <span class="ow">in</span> <span class="n">model_type</span> <span class="ow">and</span> <span class="n">gate_prior</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">gate_prior</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;vcp&quot;</span> <span class="ow">in</span> <span class="n">model_type</span> <span class="ow">and</span> <span class="n">p_capture_prior</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">p_capture_prior</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;mix&quot;</span> <span class="ow">in</span> <span class="n">model_type</span> <span class="ow">and</span> <span class="n">mixing_prior</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mixing_prior</span> <span class="o">=</span> <span class="n">jnp</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_components</span><span class="p">)</span>

    <span class="c1"># Create random key</span>
    <span class="n">rng_key</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">PRNGKey</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    
    <span class="c1"># Create model config based on model type and specified distributions</span>

    <span class="c1"># Set r distribution for model</span>
    <span class="k">if</span> <span class="n">r_dist</span> <span class="o">==</span> <span class="s2">&quot;gamma&quot;</span><span class="p">:</span>
        <span class="n">r_dist_model</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">Gamma</span><span class="p">(</span><span class="o">*</span><span class="n">r_prior</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">r_dist</span> <span class="o">==</span> <span class="s2">&quot;lognormal&quot;</span><span class="p">:</span>
        <span class="n">r_dist_model</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">LogNormal</span><span class="p">(</span><span class="o">*</span><span class="n">r_prior</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported r_dist: </span><span class="si">{</span><span class="n">r_dist</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Set r distribution for guide</span>
    <span class="k">if</span> <span class="n">r_guide</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">r_dist_guide</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">Gamma</span><span class="p">(</span><span class="o">*</span><span class="n">r_guide</span><span class="p">)</span> <span class="k">if</span> <span class="n">r_guide</span> <span class="o">==</span> <span class="s2">&quot;gamma&quot;</span> <span class="k">else</span> <span class="n">dist</span><span class="o">.</span><span class="n">LogNormal</span><span class="p">(</span><span class="o">*</span><span class="n">r_guide</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">r_dist_guide</span> <span class="o">=</span> <span class="n">r_dist_model</span>
        
    <span class="c1"># Set p distribution for model and guide</span>
    <span class="n">p_dist_model</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">Beta</span><span class="p">(</span><span class="o">*</span><span class="n">p_prior</span><span class="p">)</span>
    <span class="n">p_dist_guide</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">Beta</span><span class="p">(</span><span class="o">*</span><span class="n">p_prior</span><span class="p">)</span>

    <span class="c1"># Configure gate distribution if needed</span>
    <span class="n">gate_dist_model</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">gate_dist_guide</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="s2">&quot;zinb&quot;</span> <span class="ow">in</span> <span class="n">model_type</span><span class="p">:</span>
        <span class="n">gate_dist_model</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">Beta</span><span class="p">(</span><span class="o">*</span><span class="n">gate_prior</span><span class="p">)</span>
        <span class="n">gate_dist_guide</span> <span class="o">=</span> <span class="n">gate_dist_model</span>
            
    <span class="c1"># Configure capture probability distribution if needed</span>
    <span class="n">p_capture_dist_model</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">p_capture_dist_guide</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="s2">&quot;vcp&quot;</span> <span class="ow">in</span> <span class="n">model_type</span><span class="p">:</span>
        <span class="n">p_capture_dist_model</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">Beta</span><span class="p">(</span><span class="o">*</span><span class="n">p_capture_prior</span><span class="p">)</span>
        <span class="n">p_capture_dist_guide</span> <span class="o">=</span> <span class="n">p_capture_dist_model</span>

    <span class="c1"># Configure mixing distribution if needed</span>
    <span class="n">mixing_dist_model</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">mixing_dist_guide</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="s2">&quot;mix&quot;</span> <span class="ow">in</span> <span class="n">model_type</span><span class="p">:</span>
        <span class="n">mixing_dist_model</span> <span class="o">=</span> <span class="n">dist</span><span class="o">.</span><span class="n">Dirichlet</span><span class="p">(</span><span class="n">jnp</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">mixing_prior</span><span class="p">))</span>
        <span class="n">mixing_dist_guide</span> <span class="o">=</span> <span class="n">mixing_dist_model</span>

    <span class="c1"># Create model_config with all the configurations</span>
    <span class="n">model_config</span> <span class="o">=</span> <span class="n">ModelConfig</span><span class="p">(</span>
        <span class="n">base_model</span><span class="o">=</span><span class="n">model_type</span><span class="p">,</span>
        <span class="n">r_distribution_model</span><span class="o">=</span><span class="n">r_dist_model</span><span class="p">,</span>
        <span class="n">r_distribution_guide</span><span class="o">=</span><span class="n">r_dist_guide</span><span class="p">,</span>
        <span class="n">r_param_prior</span><span class="o">=</span><span class="n">r_prior</span><span class="p">,</span>
        <span class="n">r_param_guide</span><span class="o">=</span><span class="n">r_prior</span><span class="p">,</span>
        <span class="n">p_distribution_model</span><span class="o">=</span><span class="n">p_dist_model</span><span class="p">,</span>
        <span class="n">p_distribution_guide</span><span class="o">=</span><span class="n">p_dist_guide</span><span class="p">,</span>
        <span class="n">p_param_prior</span><span class="o">=</span><span class="n">p_prior</span><span class="p">,</span>
        <span class="n">p_param_guide</span><span class="o">=</span><span class="n">p_prior</span><span class="p">,</span>
        <span class="n">gate_distribution_model</span><span class="o">=</span><span class="n">gate_dist_model</span><span class="p">,</span>
        <span class="n">gate_distribution_guide</span><span class="o">=</span><span class="n">gate_dist_guide</span><span class="p">,</span>
        <span class="n">gate_param_prior</span><span class="o">=</span><span class="n">gate_prior</span> <span class="k">if</span> <span class="n">gate_dist_model</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">gate_param_guide</span><span class="o">=</span><span class="n">gate_prior</span> <span class="k">if</span> <span class="n">gate_dist_model</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">p_capture_distribution_model</span><span class="o">=</span><span class="n">p_capture_dist_model</span><span class="p">,</span>
        <span class="n">p_capture_distribution_guide</span><span class="o">=</span><span class="n">p_capture_dist_guide</span><span class="p">,</span>
        <span class="n">p_capture_param_prior</span><span class="o">=</span><span class="n">p_capture_prior</span> <span class="k">if</span> <span class="n">p_capture_dist_model</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">p_capture_param_guide</span><span class="o">=</span><span class="n">p_capture_prior</span> <span class="k">if</span> <span class="n">p_capture_dist_model</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">n_components</span><span class="o">=</span><span class="n">n_components</span><span class="p">,</span>
        <span class="n">mixing_distribution_model</span><span class="o">=</span><span class="n">mixing_dist_model</span><span class="p">,</span>
        <span class="n">mixing_distribution_guide</span><span class="o">=</span><span class="n">mixing_dist_guide</span><span class="p">,</span>
        <span class="n">mixing_param_prior</span><span class="o">=</span><span class="n">mixing_prior</span><span class="p">,</span>
        <span class="n">mixing_param_guide</span><span class="o">=</span><span class="n">mixing_prior</span>
    <span class="p">)</span>
    
    <span class="c1"># Get model and guide functions</span>
    <span class="n">model</span><span class="p">,</span> <span class="n">guide</span> <span class="o">=</span> <span class="n">get_model_and_guide</span><span class="p">(</span><span class="n">model_type</span><span class="p">)</span>
    
    <span class="c1"># Set up SVI</span>
    <span class="n">svi</span> <span class="o">=</span> <span class="n">SVI</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">guide</span><span class="p">,</span> <span class="n">optimizer</span><span class="p">,</span> <span class="n">loss</span><span class="o">=</span><span class="n">loss</span><span class="p">)</span>
    
    <span class="c1"># Prepare model arguments</span>
    <span class="n">model_args</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;n_cells&#39;</span><span class="p">:</span> <span class="n">n_cells</span><span class="p">,</span>
        <span class="s1">&#39;n_genes&#39;</span><span class="p">:</span> <span class="n">n_genes</span><span class="p">,</span>
        <span class="s1">&#39;counts&#39;</span><span class="p">:</span> <span class="n">count_data</span><span class="p">,</span>
        <span class="s1">&#39;batch_size&#39;</span><span class="p">:</span> <span class="n">batch_size</span><span class="p">,</span>
        <span class="s1">&#39;model_config&#39;</span><span class="p">:</span> <span class="n">model_config</span>
    <span class="p">}</span>
    
    <span class="c1"># Add total_counts for NBDM model</span>
    <span class="k">if</span> <span class="n">model_type</span> <span class="o">==</span> <span class="s2">&quot;nbdm&quot;</span><span class="p">:</span>
        <span class="n">model_args</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">count_data</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    
    <span class="c1"># Run inference</span>
    <span class="n">svi_results</span> <span class="o">=</span> <span class="n">svi</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="n">rng_key</span><span class="p">,</span>
        <span class="n">n_steps</span><span class="p">,</span>
        <span class="n">stable_update</span><span class="o">=</span><span class="n">stable_update</span><span class="p">,</span>
        <span class="o">**</span><span class="n">model_args</span>
    <span class="p">)</span>
    
    <span class="c1"># Create results object</span>
    <span class="k">if</span> <span class="n">adata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">ScribeResults</span><span class="o">.</span><span class="n">from_anndata</span><span class="p">(</span>
            <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
            <span class="n">params</span><span class="o">=</span><span class="n">svi_results</span><span class="o">.</span><span class="n">params</span><span class="p">,</span>
            <span class="n">loss_history</span><span class="o">=</span><span class="n">svi_results</span><span class="o">.</span><span class="n">losses</span><span class="p">,</span>
            <span class="n">model_type</span><span class="o">=</span><span class="n">model_type</span><span class="p">,</span>
            <span class="n">model_config</span><span class="o">=</span><span class="n">model_config</span><span class="p">,</span>
            <span class="n">n_components</span><span class="o">=</span><span class="n">n_components</span><span class="p">,</span>
            <span class="n">prior_params</span><span class="o">=</span><span class="p">{</span>
                <span class="s1">&#39;r_prior&#39;</span><span class="p">:</span> <span class="n">r_prior</span><span class="p">,</span>
                <span class="s1">&#39;p_prior&#39;</span><span class="p">:</span> <span class="n">p_prior</span><span class="p">,</span>
                <span class="s1">&#39;gate_prior&#39;</span><span class="p">:</span> <span class="n">gate_prior</span> <span class="k">if</span> <span class="s2">&quot;zinb&quot;</span> <span class="ow">in</span> <span class="n">model_type</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s1">&#39;p_capture_prior&#39;</span><span class="p">:</span> <span class="n">p_capture_prior</span> <span class="k">if</span> <span class="s2">&quot;vcp&quot;</span> <span class="ow">in</span> <span class="n">model_type</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s1">&#39;mixing_prior&#39;</span><span class="p">:</span> <span class="n">mixing_prior</span> <span class="k">if</span> <span class="n">n_components</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">ScribeResults</span><span class="p">(</span>
            <span class="n">params</span><span class="o">=</span><span class="n">svi_results</span><span class="o">.</span><span class="n">params</span><span class="p">,</span>
            <span class="n">loss_history</span><span class="o">=</span><span class="n">svi_results</span><span class="o">.</span><span class="n">losses</span><span class="p">,</span>
            <span class="n">n_cells</span><span class="o">=</span><span class="n">n_cells</span><span class="p">,</span>
            <span class="n">n_genes</span><span class="o">=</span><span class="n">n_genes</span><span class="p">,</span>
            <span class="n">model_type</span><span class="o">=</span><span class="n">model_type</span><span class="p">,</span>
            <span class="n">model_config</span><span class="o">=</span><span class="n">model_config</span><span class="p">,</span>
            <span class="n">n_components</span><span class="o">=</span><span class="n">n_components</span><span class="p">,</span>
            <span class="n">prior_params</span><span class="o">=</span><span class="p">{</span>
                <span class="s1">&#39;r_prior&#39;</span><span class="p">:</span> <span class="n">r_prior</span><span class="p">,</span>
                <span class="s1">&#39;p_prior&#39;</span><span class="p">:</span> <span class="n">p_prior</span><span class="p">,</span>
                <span class="s1">&#39;gate_prior&#39;</span><span class="p">:</span> <span class="n">gate_prior</span> <span class="k">if</span> <span class="s2">&quot;zinb&quot;</span> <span class="ow">in</span> <span class="n">model_type</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s1">&#39;p_capture_prior&#39;</span><span class="p">:</span> <span class="n">p_capture_prior</span> <span class="k">if</span> <span class="s2">&quot;vcp&quot;</span> <span class="ow">in</span> <span class="n">model_type</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s1">&#39;mixing_prior&#39;</span><span class="p">:</span> <span class="n">mixing_prior</span> <span class="k">if</span> <span class="n">n_components</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="p">}</span>
        <span class="p">)</span>
    
    <span class="k">return</span> <span class="n">results</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Manuel Razo.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>