

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>svi &mdash; SCRIBE  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="results" href="results.html" />
    <link rel="prev" title="models_mix" href="models_mix.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            SCRIBE
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickoverview.html">Quick Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Available Models</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../models/index.html">Models</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../results.html">Results Class</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="models.html">models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models_mix.html">models_mix</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">svi</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#scribe.svi.create_svi_instance"><code class="docutils literal notranslate"><span class="pre">create_svi_instance()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scribe.svi.run_inference"><code class="docutils literal notranslate"><span class="pre">run_inference()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#scribe.svi.run_scribe"><code class="docutils literal notranslate"><span class="pre">run_scribe()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="results.html">results</a></li>
<li class="toctree-l2"><a class="reference internal" href="sampling.html">sampling</a></li>
<li class="toctree-l2"><a class="reference internal" href="stats.html">stats</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html">utils</a></li>
<li class="toctree-l2"><a class="reference internal" href="viz.html">viz</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">SCRIBE</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API Reference</a></li>
      <li class="breadcrumb-item active">svi</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/svi.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-scribe.svi">
<span id="svi"></span><h1>svi<a class="headerlink" href="#module-scribe.svi" title="Link to this heading"></a></h1>
<p>Stochastic Variational Inference (SVI) module for single-cell RNA sequencing
data analysis.</p>
<p>This module implements a Dirichlet-Multinomial model for scRNA-seq count data
using Numpyro for variational inference.</p>
<dl class="py function">
<dt class="sig sig-object py" id="scribe.svi.create_svi_instance">
<span class="sig-prename descclassname"><span class="pre">scribe.svi.</span></span><span class="sig-name descname"><span class="pre">create_svi_instance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model_type=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer=&lt;numpyro.optim.Adam</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loss=&lt;numpyro.infer.elbo.TraceMeanField_ELBO</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scribe/svi.html#create_svi_instance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scribe.svi.create_svi_instance" title="Link to this definition"></a></dt>
<dd><p>Create an SVI instance with the defined model and guide.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>optional</em>) – <dl class="simple">
<dt>Type of model to use. Must be one of:</dt><dd><ul>
<li><p>”nbdm”: Negative Binomial model</p></li>
<li><p>”zinb”: Zero-Inflated Negative Binomial model</p></li>
<li><p>”nbvcp”: Negative Binomial with variable capture probability</p></li>
<li><p>”zinbvcp”: Zero-Inflated Negative Binomial with variable capture
probability</p></li>
<li><p>Mixture variants with “_mix” suffix (e.g. “nbdm_mix”)</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>optimizer</strong> (<em>numpyro.optim.optimizers</em><em>, </em><em>optional</em>) – Optimizer to use for stochastic optimization. Default is Adam with
step_size=0.001</p></li>
<li><p><strong>loss</strong> (<em>numpyro.infer.elbo</em><em>, </em><em>optional</em>) – Loss function for variational inference. Default is TraceMeanField_ELBO()</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Configured SVI instance ready for training</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpyro.infer.SVI</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.13)"><strong>ValueError</strong></a> – If model_type is invalid</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scribe.svi.run_inference">
<span class="sig-prename descclassname"><span class="pre">scribe.svi.</span></span><span class="sig-name descname"><span class="pre">run_inference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">svi_instance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nbdm'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stable_update</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scribe/svi.html#run_inference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scribe.svi.run_inference" title="Link to this definition"></a></dt>
<dd><p>Run stochastic variational inference on the provided count data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>svi_instance</strong> (<em>numpyro.infer.SVI</em>) – Configured SVI instance for running inference. NOTE: Make sure that this
was created with the same model and guide as the one being used here!</p></li>
<li><p><strong>rng_key</strong> (<em>jax.random.PRNGKey</em>) – Random number generator key</p></li>
<li><p><strong>counts</strong> (<em>jax.numpy.ndarray</em>) – Count matrix. If cells_axis=0 (default), shape is (n_cells, n_genes).
If cells_axis=1, shape is (n_genes, n_cells).</p></li>
<li><p><strong>n_steps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>optional</em>) – Number of optimization steps. Default is 100,000</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>optional</em>) – Mini-batch size for stochastic optimization. Default is 512</p></li>
<li><p><strong>model_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>optional</em>) – <p>Type of model being used. Built-in options are:
- “nbdm”: Negative Binomial-Dirichlet Multinomial model
- “zinb”: Zero-Inflated Negative Binomial model
- “nbvcp”: Negative Binomial with variable capture probability
- “zinbvcp”: Zero-Inflated Negative Binomial with variable capture</p>
<blockquote>
<div><p>probability</p>
</div></blockquote>
<ul>
<li><p>”nbdm_mix”: Negative Binomial-Dirichlet Multinomial Mixture Model</p></li>
</ul>
</p></li>
<li><p><strong>n_components</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>optional</em>) – Number of components to fit for mixture models. Default is None.</p></li>
<li><p><strong>prior_params</strong> (<em>Dict</em><em>, </em><em>optional</em>) – Dictionary of prior parameters specific to the model.</p></li>
<li><p><strong>cells_axis</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>optional</em>) – Axis along which cells are arranged. 0 means cells are rows (default), 1
means cells are columns.</p></li>
<li><p><strong>custom_args</strong> (<em>Dict</em><em>, </em><em>optional</em>) – Dictionary of additional arguments to pass to any custom model.</p></li>
<li><p><strong>stable_update</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to use stable update method. Default is True. If true,  returns
the current state if the the loss or the new state contains invalid
values.</p></li>
<li><p><strong>model_config</strong> (<em>ModelConfig</em><em> | </em><em>None</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Results from the SVI run containing optimized parameters and loss
history</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpyro.infer.svi.SVIRunResult</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="scribe.svi.run_scribe">
<span class="sig-prename descclassname"><span class="pre">scribe.svi.</span></span><span class="sig-name descname"><span class="pre">run_scribe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">counts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_inflated=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_capture=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mixture_model=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_dist='gamma'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_prior=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_prior=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gate_prior=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_capture_prior=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mixing_prior=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_steps=100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer=&lt;numpyro.optim.Adam</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cells_axis=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed=42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stable_update=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r_guide=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loss=&lt;numpyro.infer.elbo.TraceMeanField_ELBO</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/scribe/svi.html#run_scribe"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#scribe.svi.run_scribe" title="Link to this definition"></a></dt>
<dd><p>Run SCRIBE inference pipeline to fit a probabilistic model to single-cell
RNA sequencing data.</p>
<p>This function provides a high-level interface to configure and run SCRIBE
models. It handles data preprocessing, model configuration, variational
inference, and results packaging.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>counts</strong> (<em>Union</em><em>[</em><em>jnp.ndarray</em><em>, </em><em>AnnData</em><em>]</em>) – Count matrix or AnnData object containing counts. If AnnData, counts
should be in .X or specified layer. Shape should be (cells, genes) if
cells_axis=0, or (genes, cells) if cells_axis=1.</p></li>
<li><p><strong>Configuration</strong> (<em>Model</em>)</p></li>
<li><p><strong>-------------------</strong></p></li>
<li><p><strong>zero_inflated</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>default=False</em>) – Whether to use zero-inflated negative binomial (ZINB) model to account
for dropout events</p></li>
<li><p><strong>variable_capture</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>default=False</em>) – Whether to model cell-specific mRNA capture efficiencies</p></li>
<li><p><strong>mixture_model</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>default=False</em>) – Whether to use mixture model components for heterogeneous populations</p></li>
<li><p><strong>n_components</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>]</em><em>, </em><em>default=None</em>) – Number of mixture components. Required if mixture_model=True.</p></li>
<li><p><strong>Choices</strong> (<em>Distribution</em>)</p></li>
<li><p><strong>-------------------</strong></p></li>
<li><p><strong>r_dist</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>default=&quot;gamma&quot;</em>) – <dl class="simple">
<dt>Distribution family for dispersion parameter. Options:</dt><dd><ul>
<li><p>”gamma”: Gamma distribution (default)</p></li>
<li><p>”lognormal”: Log-normal distribution</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>Parameters</strong> (<em>Training</em>)</p></li>
<li><p><strong>----------------</strong></p></li>
<li><p><strong>r_prior</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>]</em><em>, </em><em>default=None</em>) – <dl class="simple">
<dt>Prior parameters for dispersion (r) distribution:</dt><dd><ul>
<li><p>For gamma: (shape, rate), defaults to (2, 0.1)</p></li>
<li><p>For lognormal: (mu, sigma), defaults to (1, 1)</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>p_prior</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>]</em><em>, </em><em>default=None</em>) – Prior parameters (alpha, beta) for success probability Beta
distribution. Defaults to (1, 1).</p></li>
<li><p><strong>gate_prior</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>]</em><em>, </em><em>default=None</em>) – Prior parameters (alpha, beta) for dropout gate Beta distribution. Only
used if zero_inflated=True. Defaults to (1, 1).</p></li>
<li><p><strong>p_capture_prior</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>]</em><em>, </em><em>default=None</em>) – Prior parameters (alpha, beta) for capture efficiency Beta distribution.
Only used if variable_capture=True. Defaults to (1, 1).</p></li>
<li><p><strong>mixing_prior</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.13)"><em>tuple</em></a><em>]</em><em>, </em><em>default=None</em>) – Prior concentration parameter(s) for mixture weights Dirichlet
distribution. Required if mixture_model=True.</p></li>
<li><p><strong>Parameters</strong></p></li>
<li><p><strong>------------------</strong></p></li>
<li><p><strong>n_steps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>default=100_000</em>) – Number of optimization steps for variational inference</p></li>
<li><p><strong>batch_size</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>]</em><em>, </em><em>default=None</em>) – Mini-batch size for stochastic optimization. If None, uses full dataset.</p></li>
<li><p><strong>optimizer</strong> (<em>numpyro.optim.optimizers</em><em>, </em><em>default=Adam</em><em>(</em><em>step_size=0.001</em><em>)</em>) – Optimizer for variational inference</p></li>
<li><p><strong>Handling</strong> (<em>Data</em>)</p></li>
<li><p><strong>-------------</strong></p></li>
<li><p><strong>cells_axis</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>default=0</em>) – Axis for cells in count matrix (0=rows, 1=columns)</p></li>
<li><p><strong>layer</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em><em>, </em><em>default=None</em>) – Layer in AnnData to use for counts. If None, uses .X.</p></li>
<li><p><strong>Options</strong> (<em>Additional</em>)</p></li>
<li><p><strong>-----------------</strong></p></li>
<li><p><strong>seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>default=42</em>) – Random seed for reproducibility</p></li>
<li><p><strong>stable_update</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>default=True</em>) – Whether to use numerically stable parameter updates during optimization</p></li>
<li><p><strong>r_guide</strong> (<em>Optional</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em><em>, </em><em>default=None</em>) – Distribution family for guide of dispersion parameter. If None, uses
same as r_dist. Options: “gamma” or “lognormal”.</p></li>
<li><p><strong>loss</strong> (<em>numpyro.infer.elbo</em><em>, </em><em>default=TraceMeanField_ELBO</em><em>(</em><em>)</em>) – Loss function for variational inference</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Results object containing:</dt><dd><ul class="simple">
<li><p>Fitted model parameters</p></li>
<li><p>Loss history</p></li>
<li><p>Model configuration</p></li>
<li><p>Prior parameters</p></li>
<li><p>Dataset metadata</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="results.html#scribe.results.ScribeResults" title="scribe.results.ScribeResults">ScribeResults</a></p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="models_mix.html" class="btn btn-neutral float-left" title="models_mix" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="results.html" class="btn btn-neutral float-right" title="results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Manuel Razo.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>